### PR TITLE
a bit of cleanup, and fix the url assignment ordering

### PR DESCRIPTION
**Pull request by whyrusleeping**
This might fix our gateway issues

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -10,6 +10,30 @@ import (
...
+	req_a.Close = true
+	rt := &http.Transport{DisableKeepAlives: true}
+	resp, err := rt.RoundTrip(&req_a)
+	if err != nil {
+		log.Printf("target %s failed: %s", target, err)
+		close(fail)
+		return
+	}
+
+	if ra.StatusCode >= 500 || ra.StatusCode == 408 {
```
Comment (ghost) [inline main.go]: this should include 404 and 400 too, based on #4

File: main.go [hunk 2]
```diff
@@ -50,43 +74,8 @@ func main() {
...
-				close(fail_b)
-			} else if resp.StatusCode >= 500 || resp.StatusCode == 408 {
-				log.Printf("target B unsatisfying status: %d", resp.StatusCode)
-				close(fail_b)
-			} else {
-				resp_b <- resp
-			}
-		}()
+		go forwardRequest(r, ua, resp_a, cancel_a, fail_a)
+		go forwardRequest(r, ub, resp_b, cancel_b, fail_b)
```
Comment (ghost) [inline main.go]: this needs the cancel channels swapped -- a should cancel b, and vice versa

File: main.go [hunk 3]
```diff
@@ -10,6 +10,30 @@ import (
 	"strings"
 )
 
+func forwardRequest(r *http.Request, target *url.URL, out chan *http.Response, cancel chan struct{}, fail chan struct{}) {
+	req_a := *r
+	req_a.URL, _ = url.Parse(r.URL.String())
+	req_a.URL.Host = target.Host
+	req_a.Cancel = cancel
+
+	req_a.Close = true
```
Comment (ghost) [inline main.go:20]: explicit Close isn't needed if DisableKeepAlives is enabled

Comment (ghost) [inline main.go:20]: We might need to close the upstream response body though, once we've finished copying it into the downstream response.

>  The client must close the response body when finished with it: 

-- https://golang.org/pkg/net/http/

Comment (whyrusleeping) [inline main.go:20]: good catch, thanks!

File: main.go (hunk 1 above)
Comment (whyrusleeping) [inline main.go]: i'm not so sure thats right though...

File: main.go (hunk 2 above)
Comment (whyrusleeping) [inline main.go]: again, good catch :+1:

Comment (ghost) [inline main.go]: > this needs the cancel channels swapped -- a should cancel b, and vice versa

sorry nevermind, it's correct

Comment (whyrusleeping) [inline main.go]: er, wait. no. `cancel_a` is the channel that cancels request a, so this is correct.

**Comment by ghost** [comment]
LGTM :+1:

**Pull request merged**